src/problem.jl

```diff
@@ -0,0 +1,82 @@
+"""
+    StateSpaceProblem{isinplace, dType, uType,tType,noiseType,oType, lType} <: AbstractStateSpaceProblem{isinplace}
+
+A state space problem contains all the information necessary to simulate a state space
+process.
+"""
+struct StateSpaceProblem{isinplace, dType, uType,tType,noiseType,oType, lType} <: AbstractStateSpaceProblem{isinplace}
+    f::Function # Evolution
+    g::Function # Noise function
+    h::Function # Observation function
+    D::dType # Noise distribution
+    u0::uType # Initial condition
+    tspan::tType # Timespan to use
+    noise::noiseType # The noise to use
+    observables::oType # Data to use, if any
+    likelihood::lType # Likelihood function, if any
+end
+
+"""
+    StateSpaceProblem(f, g, h, D, u0, tspan, noise, observables, likelihood=((x...) -> 0.0))
+"""
+function StateSpaceProblem(f, g, h, D, u0, tspan, noise, observables, likelihood=((x...) -> 0.0))
+    return StateSpaceProblem{
+        Val(false), # Default to out-of-place
+        typeof(D),
+        typeof(u0),
+        typeof(tspan),
+        typeof(noise),
+        typeof(observables),
+        typeof(likelihood)
+    }(f, g, h, D, u0, SciMLBase.promote_tspan(tspan), noise, observables, likelihood)
+end
+
+"""
+    index_or_nothing(x::AbstractArray, i::Int)
+    index_or_nothing(x::Nothing, i::Int)
+
+Extracts index `i` from `x`, if `x` is a subtype of `AbstractArray`. Otherwise,
+this function returns `nothing`.
+"""
+index_or_nothing(x::AbstractArray, i::Int) = x[i]
+index_or_nothing(x::Nothing, i::Int) = nothing
+
+"""
+    transition(prob::StateSpaceProblem, u, params, t)
+
+Calculates the next latent state, given by `f(u[t-1], params, t-1) + g(u[t-1], params, t-1)`.
+"""
+function transition(prob::StateSpaceProblem, u, params, t)
+    return prob.f(u[t - 1], params, t-1) .+ prob.g(u[t - 1], params, t-1, index_or_nothing(prob.noise, t-1))
+end
+
+"""
+    _solve(prob::StateSpaceProblem, params)
+
+Returns a `StateSpaceSolution` using the provided parameters `params`.
+Calculates the likelihood if the likelihood function is available.
+"""
+function _solve(prob::StateSpaceProblem, params)
+    t0, T = prob.tspan
+
+    # Initial noise
+    z0 = prob.h(prob.u0, params, 1, nothing)
+
+    # Preallocate arrays
+    u = Vector{typeof(prob.u0)}(undef, T+1) # Should be square, not vectored?
+    z = Vector{typeof(z0)}(undef, T+1) # Should be square, not vectored?
+
+    u[1] = prob.u0
+    z[1] = z0
+
+    # Accumulate likelihood
+    loglik = 0.0
```

Comment:
So the likelihood function is supposed to return `Float64` (otherwise this will be type unstable)? It would be good to document this or change the implementation (I was already wondering about it above when you define the default likelihood as `(x...) -> 0.0` (BTW in Julia >= 1.7 you can use `Returns(0.0)`)).

Reply:
Yeah, I don't particularly care for this implementation, especially if you have no interest in evaluating the likelihood (in which case I'd prefer `nothing` or `missing`). Perhaps this is a trait or dispatch fix where we run the likelihood function only if `!isnothing(prob.likelihood)` or something.

Reply:
The design discussed previously was that the likelhood should be clear from the algorithm choice, not a function passed in.

Furthermore, the noise distribution/etc. and whether the noise is conditioned on should determine the types of likelhoods available, right?

Reply:
For example, in https://github.com/HighDimensionalEconLab/DifferentiableStateSpaceModels.jl/blob/main/docs/src/state_space_examples.md#example-4-likelihood-and-gradient-given-noise  the `D` is a distribution as specified in https://github.com/HighDimensionalEconLab/DifferentiableStateSpaceModels.jl/blob/main/docs/src/state_space_examples.md#example-1-general-state-space-model   so I don't see why the likelihood isn't explicit if conditioned on the noise.

Reply:
I agree. Dispatched types on algorithms will be added later on.

Reply:
Though I will say my preference is that I would like people to be able to pass in their likelihoods if they wish -- trying too hard to infer from `D` is a fine idea but it reads as a little too paternalistic. These are all interface methods and they should be very, very general and customizable.

Reply:
There are a few issues with that.  The most practical is that it means that some sort of rrule needs to go through their likelihood closure to the underlying distributions that they create.  Maybe that is easy to write and doesn't look weird from 
 auser perspective, and maybe it is overhead free, but I never trust zygote.

But the bigger question I have is:  leave aside the special cases for the linear, etc. or things where we agree that you have a specialized algorithm.  And remember that if you are implementing an unscented kalman filter/etc. it is in a new "algorithm".  Otherwise, is there any variation at all in the calculation of the likelihood given this specification of the model when either
1. Things are unobservable with a `D` , but you are conditioning on the `noise`
2. Things are perfectly observable (i.e. D = nothing) but the `noise` itself is given as a distribution?

In both of those cases, isn't there only a single way to do the likelihood?

Reply:
For our case, sure, but it should be overrideable as a default. We can provide convenient defaults but should never restrict the likelihood function to something. 

As to the `rrule`, it isn't our responsibility to make sure the the AD works for a custom likelihood. That's on the user side. For defaults, absolutely we need to guarantee that AD does the correct thing every time. 

These "defaults" I'm referring to are not in this PR but will be later.